[FLINK-11514][client] Port ClusterClientTest

zentol opened the pull request:
<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Travis CI to do that following [this guide](http://flink.apache.org/contribute-code.html#best-practices).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change

This PR ports the `ClusterClientTest` to not use legacy code

## Brief change log

Removed:

#testClusterClientShutdown
* Irrelevant for RestClusterClient

#testClusterClientStop/#testClusterClientCancel
* Subsumed by RestClusterClient#testJobSubmitCancelStop

#testClusterClientSavepoint
* Subsumed by RestClusterClient#testTriggerSavepoint

#testClusterClientList
* Subsumed in RestClusterClientTest#testListJobs

#testDisposeSavepointUnknownResponse
* Invalid case under new REST framework, impossible under usual circumstances. (can only happen iff the client/server disagree about headers, but they are versioned and stable)

#testDisposeClassNotFoundException
* RestClusterClient is not responsible for enriching error messages.

Ported:

#testClusterClientCancelWithSavepoint
* new test RestClusterClientTest#testTriggerCancelWithSavepoint

GJL commented on flink-clients/src/test/java/org/apache/flink/client/program/rest/RestClusterClientTest.java:
```diff
@@ -448,20 +457,24 @@ public void testTriggerSavepoint() throws Exception {
 				{
-					CompletableFuture<String> savepointPathFuture = restClusterClient.triggerSavepoint(id, null);
-					String savepointPath = savepointPathFuture.get();
+					String savepointPath = cancelJob
```
If we only assert on the job cancel flag being set to `true`, a single test would be enough imo:

```
       @Test
	public void testCancelJobWithSavepointShouldTriggerSavepointWithCancelJobFlag() throws Exception {
		final boolean cancelJob = true;
		final String expectedSavepointLocation = "/savepoints/savepoint-0d2fb9-8d5e0106041a";

		final TestSavepointHandlers testSavepointHandlers = new TestSavepointHandlers();
		final TestSavepointHandlers.TestSavepointTriggerHandler triggerHandler =
			testSavepointHandlers.new TestSavepointTriggerHandler(
				cancelJob, "/savepoints");
		final TestSavepointHandlers.TestSavepointHandler savepointHandler =
			testSavepointHandlers.new TestSavepointHandler(new SavepointInfo(
				expectedSavepointLocation,
				null));

		try (TestRestServerEndpoint restServerEndpoint = createRestServerEndpoint(
			triggerHandler,
			savepointHandler)) {
			
			final RestClusterClient<?> restClusterClient = createRestClusterClient(restServerEndpoint.getServerAddress().getPort());
			final String savepointLocation = restClusterClient.cancelWithSavepoint(new JobID(), "/savepoints");
			assertEquals(expectedSavepointLocation, savepointLocation);
		}
	}
```

It was a mistake to have that many scenarios + asserts in `testTriggerSavepoint`. The test was hard to read to begin with and this change makes it even harder. 

If you think the current state is readable enough, maybe consider eliminating the code duplication:
```
	private static String triggerSavepoint(final boolean cancelJob, final String targetSavepointDirectory, final RestClusterClient<?> restClusterClient) throws Exception {
		return cancelJob
			? restClusterClient.cancelWithSavepoint(new JobID(), targetSavepointDirectory)
			: restClusterClient.triggerSavepoint(new JobID(), targetSavepointDirectory).get();
	}
```

zentol replied:
We have to test both scenarious (cancel being true and false), as otherwise we wouldn't detect bugs where the flag is _always_ set or parsed to true of false.

zentol replied:
I agree that the test is overloaded, but don't think this flag makes this significantly worse. I would be up for refactoring the test into 4-5 separate ones in a follow-up.

GJL replied:
I meant that a single test for testing that the flag can be set to `true` would be enough. For the `false` case all scenarios in `testTriggerSavepoint` should be kept.

zentol replied:
that's fair.

GJL commented on flink-clients/src/test/java/org/apache/flink/client/program/rest/RestClusterClientTest.java:
```diff
@@ -448,20 +457,24 @@ public void testTriggerSavepoint() throws Exception {
 				{
 					try {
-						restClusterClient.triggerSavepoint(id, null).get();
+						String savepointPath = cancelJob
```
I wouldn't assign to an unused variable `savepointPath` here.

GJL commented:
@flinkbot approve description
@flinkbot approve consensus

zentol commented:
@GJL I've addressed your comments, and then decided to rework the savepoint-trigger tests.
Additionally I fixed a bug in the RestClusterClientTest where we never actually caused a failure in the client since the wrong method was overridden.

GJL commented:
Thanks, I will have a look tomorrow.

GJL commented on flink-clients/src/test/java/org/apache/flink/client/program/rest/RestClusterClientSavepointTriggerTest.java:
```diff
@@ -0,0 +1,340 @@
+		return testRestServerEndpoint;
+	}
+
+	private static class TestRestServerEndpoint extends RestServerEndpoint implements AutoCloseable {
```
I would extract this class to avoid code duplication with the other `RestClusterClientTest`. Same with `TestHandler`.

GJL commented on flink-clients/src/test/java/org/apache/flink/client/program/rest/RestClusterClientSavepointTriggerTest.java:
```diff
@@ -0,0 +1,340 @@
+		config.setLong(RestOptions.RETRY_DELAY, 0);
+		config.setInteger(RestOptions.PORT, 0);
+
+		restConfig = config;
```
I would use a `UnmodifiableConfiguration` to avoid accidental modifications from tests. Also, by convention`static finals` should be uppercased: https://docs.oracle.com/javase/tutorial/java/nutsandbolts/variables.html

GJL commented on flink-clients/src/test/java/org/apache/flink/client/program/rest/RestClusterClientSavepointTriggerTest.java:
```diff
@@ -0,0 +1,340 @@
+
+	@BeforeClass
+	public static void setUp() throws ConfigurationException {
+		restServerEndpointConfiguration = RestServerEndpointConfiguration.fromConfiguration(restConfig);
```
`restServerEndpointConfiguration` could be as well `static final` and initialized in the static initializer.

zentol replied:
One could, but `RestServerEndpointConfiguration.fromConfiguration` may throw exception and doing so in the static blocks leads to misleading stacktraces and odd behaviors.

If this call were to fail the first time the test is run you'll get a ExceptionOnInitializationError (i.e. the classes cannot be loaded), but the JVM doesn't try to load the class again resulting in, iirc, ClassNotFoundExceptions or something.

GJL replied:
ok

zentol commented:
@GJL I've addressed your comments.

GJL commented:
@zentol Looks good.

GJL commented:
@flinkbot approve all
